### DISCUSSION TITLE
[ESLint rule idea] secure-browser-globals

### DISCUSSION THREAD

**Discussion by ljosberinn** [vercel/next.js#23479] [Ideas]
### Describe the feature you'd like to request

Next.js prerenders outside of the browser context and as such, `window` and `document` are undefined outside of `use(Layout)Effect`. This is a "limitation" many users including myself trip(ped) over and is part of the usual learning curve of Next.js or any serverside JS lib/framework.

### Describe the solution you'd like

`@next/eslint-plugin` could provide a linting rule that checks whether these criteria are met when accessing `window.` or `document.`:
- access does not happen on the top level of a module
```js
// error
const supportsMatchMedia = 'matchMedia' in window;

export function MyComponent() {
  return null;
}
```
- access does not happen on the top level of a component:
```js
export function MyComponent() {
  // error
  const supportsMatchMedia = 'matchMedia' in window;

  return null;
}
```
- access does not happen in `getStaticProps`/`getServerSideProps`/`getStaticPaths`:
```js
export default function MyPage(props) {
  return null;
}

export async function getStaticProps(context) {
  // error
  const supportsMatchMedia = 'matchMedia' in window;

  return {
    props: { supportsMatchMedia }
  }
}
```

Limitations:
- access within functions that are called within `useEffect` or any of the above `getXYProps` may not be validated as we don't know their execution context:
```js
import { useEffect } from 'react';

// could be save, could also not be safe :shrug:
function doSomethingOnWindow() {
  console.log(window.matchMedia)
}

export function MyComponent() {
  useEffect(() => {
    doSomethingOnWindow()
  }, [])

  return null;
}
```

### Describe alternatives you've considered

- turning off `browser` environment within the ESLint config
- `no-restricted-globals`, which would require a list of properties, and even then would unconditionally report them

**Comment by housseindjirdeh** [comment]
Definitely a great idea! If you (or anyone) would like to submit a PR to add this now that Next.js 11 supports ESLint out-of-the-box - that would be great :). Would definitely be useful for folks 🚀
